src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -382,30 +383,32 @@ static long pickChunkOffset(long chunkOffset, long byteWidth, int chunkWidth) {
                 : chunkOffset;
     }
 
-    // Minimum allocation size = maximum cached size
-    private static final int CACHED_BUFFER_SIZE = 256;
-
     @ForceInline
     @SuppressWarnings("restricted")
     public static Arena newBoundedArena(long size) {
-        // JDK-8347997: buffer cache pinned section needs to happen outside of constructor.
-        long bufferSize = Math.max(size, CACHED_BUFFER_SIZE);
-        long fromCache = bufferSize == CACHED_BUFFER_SIZE ? CallBufferCache.acquire() : 0;
-        long address = fromCache != 0 ? fromCache : CallBufferCache.allocate(bufferSize);
-        return new BoundedArena(MemorySegment.ofAddress(address).reinterpret(size));
+        // JDK-8347997: buffer cache pinned section needs to happen outside of constructor and before
+        // confined session, otherwise scalar replacement breaks.
+        MemorySegment unscoped = CallBufferCache.acquireOrAllocate(size);
+        Arena scope = Arena.ofConfined();
+        MemorySegment source = unscoped.reinterpret(scope, null);
```

Comment:
I suggest passing the `scope` to `acquireOrAllocate` as well, so we only need a single call to `reinterpret` (it has an overload that takes both a size and arena.

Reply:
I generally agree that would be the right thing to do. But again, this would require creating the scope before the CTL interaction and that breaks EA. Just swapping lines 391 and 392 leads to plenty of allocations.